### PR TITLE
AK: Extend error propagation in ByteBuffer

### PR DESCRIPTION
**Pull request by RasmusNylander**
This makes creating, sliceing, copying, and overwriting propagate errors using ErrorOr instead of just panicing and removes any non error propagating variants of error propagating functions.

This also adds the abillity to overwrite using a ReadonlyBytes instead of a void const* and size_t which is quite nice.

### CODE REVIEW

**Comment by RasmusNylander** [comment]
Ups! I forgot to fix the tests. It will have to wait until tomorrow

File: Kernel/Storage/StorageDevice.cpp [hunk 1]
```diff
@@ -133,7 +130,7 @@ ErrorOr<size_t> StorageDevice::write(OpenFileDescription&, u64 offset, const Use
     // then write the whole block back to the disk.
     if (remaining > 0) {
         // FIXME: Do something sensible with this OOM scenario.
```
Comment (alimpfard) [inline Kernel/Storage/StorageDevice.cpp:132]: Please also remove this fixme.

File: Userland/Libraries/LibCore/Stream.h [hunk 2]
```diff
@@ -527,11 +527,9 @@ class BufferedHelper {
         if (!stream.is_open())
             return ENOTCONN;
 
-        auto maybe_buffer = ByteBuffer::create_uninitialized(buffer_size);
-        if (!maybe_buffer.has_value())
-            return ENOMEM;
+        auto buffer = TRY(ByteBuffer::create_uninitialized(buffer_size));
 
-        return BufferedType<T> { move(stream), maybe_buffer.release_value() };
+        return BufferedType<T> { move(stream), buffer };
```
Comment (alimpfard) [inline Userland/Libraries/LibCore/Stream.h:532]: Move this buffer in.
I'm pretty sure the copy ctor actually duplicates the buffer

File: AK/ByteBuffer.h [hunk 3]
```diff
@@ -61,31 +61,32 @@ class ByteBuffer {
...
 
-    [[nodiscard]] static Optional<ByteBuffer> create_uninitialized(size_t size)
+    [[nodiscard]] static ErrorOr<ByteBuffer> create_uninitialized(size_t size)
     {
         auto buffer = ByteBuffer();
-        if (buffer.try_resize(size).is_error())
-            return {};
-        return { move(buffer) };
+        TRY(buffer.try_resize(size));
+        return buffer;
```
Comment (alimpfard) [inline AK/ByteBuffer.h:69]: Some of these might end up copying, I'm not too sure right now
I'd suggest taking a look (maybe delete the copy ctor entirely to check), and replacing them with `{ move(buffer) }` where needed.

File: Userland/Libraries/LibWeb/Crypto/SubtleCrypto.cpp [hunk 4]
```diff
@@ -63,14 +63,15 @@ JS::Promise* SubtleCrypto::digest(String const& algorithm, JS::Handle<JS::Object
...
+    auto result_buffer_or_error = ByteBuffer::copy(digest.immutable_data(), hash.digest_size());
+    if (result_buffer_or_error.is_error()) {
         auto* error = wrap(wrapper()->global_object(), DOM::OperationError::create("Failed to create result buffer"));
         promise->reject(error);
         return promise;
     }
+    auto result_buffer = result_buffer_or_error.release_value();
 
-    auto* result = JS::ArrayBuffer::create(global_object, result_buffer.release_value());
+    auto* result = JS::ArrayBuffer::create(global_object, result_buffer);
```
Comment (alimpfard) [inline Userland/Libraries/LibWeb/Crypto/SubtleCrypto.cpp:74]: Move this buffer in.
(I'm sure this has happened a few times before that I've missed as well as after this commit, please take a good look)

File: AK/ByteBuffer.h [hunk 5]
```diff
@@ -135,13 +135,13 @@ class ByteBuffer {
...
-    [[nodiscard]] ByteBuffer slice(size_t offset, size_t size) const
+    [[nodiscard]] ErrorOr<ByteBuffer> slice(size_t offset, size_t size) const
     {
         // I cannot hand you a slice I don't have
-        VERIFY(offset + size <= this->size());
+        if (offset + size > this->size())
+            return Error::from_string_literal("ByteBuffer: slice out of bounds");
 
-        return MUST(copy(offset_pointer(offset), size));
+        return TRY(copy(offset_pointer(offset), size));
```
Comment (alimpfard) [inline AK/ByteBuffer.h:144]: `return TRY(X)` as the last statement is kinda meaningless.
Just `return X`

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.h [hunk 6]
```diff
@@ -212,9 +212,9 @@ Value ArrayBuffer::get_modify_set_value(size_t byte_index, Value value, ReadWrit
 
     // FIXME: Check for shared buffer
 
-    auto raw_bytes_read = buffer_impl().slice(byte_index, sizeof(T));
+    auto raw_bytes_read = MUST(buffer_impl().slice(byte_index, sizeof(T))); // FIXME: Handle error
     auto raw_bytes_modified = operation(raw_bytes_read, raw_bytes);
-    raw_bytes_modified.span().copy_to(buffer_impl().span().slice(byte_index));
+    raw_bytes_modified.span().copy_to(buffer_impl().span().slice(byte_index)); // FIXME: Handle error
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.h:217]: What error?

File: Userland/Libraries/LibTLS/Socket.cpp [hunk 7]
```diff
@@ -28,13 +28,23 @@ Optional<ByteBuffer> TLSv12::read()
...
+    if (buffer_or_error.is_error()) // FIXME: Propagate error with ErrorOr.
+        return {};
+    auto buffer = buffer_or_error.release_value();
+
+    auto new_application_buffer_or_error = m_context.application_buffer.slice(length, m_context.application_buffer.size() - length);
+    if (new_application_buffer_or_error.is_error()) // FIXME: Propagate error with ErrorOr.
+        return {};
+    auto new_application_buffer = new_application_buffer_or_error.release_value();
+
+    m_context.application_buffer = new_application_buffer;
```
Comment (alimpfard) [inline Userland/Libraries/LibTLS/Socket.cpp:46]: Again, move these in.

File: Userland/Libraries/LibTLS/Socket.cpp [hunk 8]
```diff
@@ -28,13 +28,23 @@ Optional<ByteBuffer> TLSv12::read()
...
+    auto length = min(m_context.application_buffer.size(), max_size);
+
+    auto buffer_or_error = m_context.application_buffer.slice(0, length);
+    if (buffer_or_error.is_error()) // FIXME: Propagate error with ErrorOr.
+        return {};
+    auto buffer = buffer_or_error.release_value();
+
+    auto new_application_buffer_or_error = m_context.application_buffer.slice(length, m_context.application_buffer.size() - length);
+    if (new_application_buffer_or_error.is_error()) // FIXME: Propagate error with ErrorOr.
+        return {};
```
Comment (alimpfard) [inline Userland/Libraries/LibTLS/Socket.cpp:43]: It's better to VERIFY in such cases until the user handles it correctly
Please just use `MUST()` here with the usual comment.

File: Userland/Libraries/LibTLS/Socket.cpp [hunk 9]
```diff
@@ -52,7 +62,13 @@ String TLSv12::read_line(size_t max_size)
...
 
     String line { bit_cast<char const*>(start), offset, Chomp };
-    m_context.application_buffer = m_context.application_buffer.slice(offset + 1, m_context.application_buffer.size() - offset - 1);
+
+    auto new_application_buffer_or_error = m_context.application_buffer.slice(offset + 1, m_context.application_buffer.size() - offset - 1);
+    if (new_application_buffer_or_error.is_error()) // FIXME: Propagate error with ErrorOr.
+        return {};
+    auto new_application_buffer = new_application_buffer_or_error.release_value();
+
+    m_context.application_buffer = new_application_buffer;
```
Comment (alimpfard) [inline Userland/Libraries/LibTLS/Socket.cpp:71]: Same here

File: AK/ByteBuffer.h [hunk 10]
```diff
@@ -153,11 +153,6 @@ class ByteBuffer {
         m_size = 0;
     }
 
-    ALWAYS_INLINE void resize(size_t new_size)
-    {
-        MUST(try_resize(new_size));
-    }
-
```
Comment (alimpfard) [inline AK/ByteBuffer.h:161]: Eh
This feels pointless, it's the same thing with the callsite getting uglier.
Maybe renaming the function to something like `resize_but_fixme_should_propagate_errors()` or whatever the other APIs named like that are would be better?

File: AK/ByteBuffer.h [hunk 11]
```diff
@@ -225,9 +225,9 @@ class ByteBuffer {
         other.m_inline = true;
     }
 
-    void trim(size_t size, bool may_discard_existing_data)
+    void trim_unchecked(size_t size, bool may_discard_existing_data)
     {
-        VERIFY(size <= m_size);
+        // VERIFY(size <= m_size);
```
Comment (alimpfard) [inline AK/ByteBuffer.h:238]: Please don't keep commented code.

Comment (alimpfard) [reply AK/ByteBuffer.h:238]: What's the point of this anyway?
A random VERIFY() won't really hurt anything

File: AK/ByteBuffer.h [hunk 12]
```diff
@@ -192,11 +192,14 @@ class ByteBuffer {
         MUST(append(other.data(), other.size()));
     }
 
-    void overwrite(size_t offset, void const* data, size_t data_size)
+    ErrorOr<void> overwrite(size_t offset, void const* data, size_t data_size)
     {
         // make sure we're not told to write past the end
-        VERIFY(offset + data_size <= size());
+        if (offset + data_size > size())
+            return Error::from_string_literal("ByteBuffer::overwrite: out of bounds");
```
Comment (alimpfard) [inline AK/ByteBuffer.h:199]: No.
This is not a simple runtime error, this is a hard logical error.
Please keep this as a VERIFY.

**Review by alimpfard — CHANGES_REQUESTED** [review]

File: AK/ByteBuffer.h (hunk 12 above)
Comment (RasmusNylander) [reply AK/ByteBuffer.h:199]: You're right, this is pretty stupid - I honestly don't really know what I was thinking. I'll take a look at your suggested changes later today, but I'm thinking this PR might just need to be axed.

Comment (RasmusNylander) [reply AK/ByteBuffer.h:199]: I think I will just close this pull request. I'm sorry to have wasted your time

**Pull request closed without merging**